dali/kernels/reduce/reduce_setup_utils.h

```diff
@@ -0,0 +1,194 @@
+      prev_reduced = is_reduced;
+    }
+    dim_groups.push_back(std::make_pair(group_start, group_size));
+  } else {
+    // totally degenerate
+    dim_groups.push_back(std::make_pair(0, d));
+  }
+}
+
+
+inline void CheckBatchReduce(const TensorListShape<> &tls, span<const int> axes) {
+  if (tls.num_samples() == 0)
+    return;
+
+  uint64_t mask = to_bit_mask(axes);
+  SmallVector<int, 6> non_reduced;
+  for (int a = 0; a < tls.sample_dim(); a++) {
+    if (!(mask & (1 << a)))
+      non_reduced.push_back(a);
+  }
+
+  auto first_sample_shape = tls.tensor_shape_span(0);
+
+  for (int i = 1; i < tls.num_samples(); i++) {
+    auto sample_shape = tls.tensor_shape_span(i);
+    for (int a : non_reduced) {
+      DALI_ENFORCE(sample_shape[a] == first_sample_shape[a], make_string(
+        "Reduce: batch reduction requires that all samples have the same extent in non-reduced "
+        "dimensions.\nError at sample ", i, " axis ", a, ": the extent is ", sample_shape[a],
+        " != ", first_sample_shape[a], " in the first sample in the batch."));
+    }
+  }
+}
+
+
+inline void CheckAxes(span<const int> axes, int dim) {
+  uint64_t mask = 0;
+  for (auto a : axes) {
+    if (a < 0 || a >= dim)
+      throw std::out_of_range(make_string("Axis index out of range: ", a, " not in 0..", dim-1));
+    uint64_t amask = 1ul << a;
+    if (mask & amask)
+      throw std::invalid_argument(make_string("Duplicate axis index ", a));
+    mask |= amask;
+  }
+}
+
+
+inline void CalculateReducedShape(TensorListShape<> &out_shape,
+                                  const TensorListShape<> &in_shape,
+                                  span<const int> axes,
+                                  bool keep_dims,
+                                  bool batch_reduce) {
+  int nsamples = in_shape.num_samples();
+  int out_samples = batch_reduce ? 1 : nsamples;
+  int in_dim = in_shape.sample_dim();
+  uint64_t mask = to_bit_mask(axes);
+
+  int out_dim = keep_dims ? in_dim : in_dim - axes.size();
+  out_shape.resize(out_samples, out_dim);
+  for (int i = 0; i < out_samples; i++) {
+    auto in_sample_shape = in_shape.tensor_shape_span(i);
+    auto out_sample_shape = out_shape.tensor_shape_span(i);
+    int out_d = 0;
+    for (int d = 0; d < in_dim; d++) {
+      if (mask & (1ul << d)) {
+        if (keep_dims)
+          out_sample_shape[out_d++] = 1;
+        continue;  // skip reduced axes
+      }
+      assert(out_d < out_dim);
+      out_sample_shape[out_d++] = in_sample_shape[d];
+    }
+    assert(out_d == out_dim);
+  }
+}
+
+template <typename ArrayLike, int ndim>
+inline void CalculateReductionFactors(ArrayLike &out, const TensorListShape<ndim> &in_shape,
+                                      span<const int> axes) {
+  for (int i = 0; i < in_shape.num_samples(); i++) {
+    auto sample_shape = in_shape.tensor_shape_span(i);
+    if (volume(sample_shape) == 0) {
+      out[i] = 0;  // degenerate sample - no reduction possible
+      continue;
+    }
```

Comment:
Hmmm, I'm lost in the 0-dim specialization. Isn't, in case of 0-dim, reduction also impossible, but `volume==1`?

Reply:
It's not 0-dim. It's 0-volume, regardless of number of dimensions. Still, thanks for getting my attention to this, because it's actually a bug - it should be `max<decltype(Out[0])>` not zero - the reduction is infinite.

Reply:
I know it's 0-volume. But when we have 0-dim (a scalar), reduction is also impossible, isn't it? So I suppose this condition should look something like
```
if (volume(sample_shape) == 0 || sample_shape.ndims() == 0)
```

Reply:
It's not like it. When reducing, you force a reduced dimension to 1. If the tensor is empty (0 volume) you can't conjure up the data to be reduced.
A scalar, on the other hand, can't be reduced due to a totally different condition - there's no valid axis index.
I wonder what I should do with these 0-sized tensors - when not all axes are reduced and the output is produced with `keep_dims=true`, I can still set the reduced extents to 1 and non-reduced ones to 0, so the output is empty.